### DISCUSSION TITLE
Question for non-standard force layout model

### DISCUSSION THREAD

**Discussion by NicolasWinsten** [gampleman/elm-visualization#157] [Q&A] [answered]
Hello

I am trying to make a force layout simulation that displays the graph in a tree like way.  By "tree-like", I mean I want the resulting layout to look something like a hierarchical layout where parent nodes will be displayed above child nodes, left child nodes are placed to the left of their sibling nodes, etc.  These aren't hard constraints, but I want the layout to trend towards local placements close to this:

```
          parent
        /        \
leftChild    rightChild
```

I don't want to use Sugiyama because it is complicated to turn that into an incremental algorithm (nodes in my visualization will be continually added and removed).

My main idea right now is to coarsen the graph by treating some of the local 3-node subgraphs (parent, leftChild, rightChild) as single nodes, and applying a force simulation on that simplified graph.  Maybe even treating 2-node subgraphs of (parent, leftChild) as single nodes would be enough.

Is there a more elegant/simple approach?

Thank you

**Comment by gampleman** [comment, marked as answer]
I'm not sure why the Hierarchy module doesn't fit your needs?

The main way you can mess with force layouts is that after each tick you can change the positions and velocities manually. So you could enforce further constraints, such as making a parent node have a lower y value than a child node for instance.

Another trick worth keeping in mind is that links in a force simulation don't have to correspond 1-1 with links that you draw on the screen. So for instance adding an extra invisible link between `leftChild` and `rightChild` will make it quite likely that you will end up with a triangular looking layout between the three.

**Reply by NicolasWinsten** [reply]
Thanks for the response!

Hierarchy doesn't fit perfectly because my data is not a tree, it's a graph.  I just want the layout to show some local hierarchy when possible.  I will play around with your ideas

**Reply by gampleman** [reply]
Other possibility you could try is to remove edges in your graph until it forms a tree, use Hierarchy to do layout, then use that as the initial positions for your force simulation.

**Reply by NicolasWinsten** [reply]
That ended up being the best solution.  I used Hierarchy to compute the placements of the main tree of the graph, and then applied those placements in the force simulation using `towardsX` and `towardsY`.

Thank you
